api/server/middleware_auth.go

```diff
@@ -38,6 +38,76 @@ func NewAuthMiddleware() *authMiddleware {
 type authMiddleware struct {
 }
 
+// newSecurityMiddleware based on auth configuration returns securityHandler or just
+func newSecurityMiddleware(authenticators map[string]auth.Authenticator) func(next http.HandlerFunc) http.HandlerFunc {
+	if auth.Enabled() {
+		return func(next http.HandlerFunc) http.HandlerFunc {
+			return securityHandler(authenticators, next)
+		}
+	}
+
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return next
+	}
+}
+
+// securityHandler implements Authentication and Authorization check at the same time
+// this functionality where not moved to separate functions because of simplicity
+func securityHandler(authenticators map[string]auth.Authenticator, next http.HandlerFunc) http.HandlerFunc {
+	if authenticators == nil {
+		return next
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		tokenHeader := r.Header.Get("Authorization")
+		tokens := strings.Split(tokenHeader, " ")
+
+		if len(tokens) < 2 {
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		token := tokens[1]
+
+		// Determine issuer
+		issuer, err := auth.TokenIssuer(token)
+		if err != nil {
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+
+		// Authenticate user
+		ctx := r.Context()
+		var claims *auth.Claims
+		if authenticator, exists := authenticators[issuer]; exists {
+			claims, err = authenticator.AuthenticateToken(ctx, token)
```

Comment:
Here ctx is expected to have the token inside in the metadata. You will need to save the token from the Header and place it in the metadata. See https://github.com/libopenstorage/openstorage/blob/master/api/server/volume.go#L83-L100 . You may need to create a util call now that there are two callers.

Reply:
Yes, I see. But it looks like out of scope of this particular pull request. I was aimed to secure rest calls. But bypassing this token to grpc(call by sdk) is another pull request.

https://github.com/libopenstorage/openstorage/pull/1329/files#diff-4f77727792ad8363ad8983d663ff1d72R15

Reply:
So actual extracting of token and passing it downstream to grpc call will be done in handler itself. Do you think is it a bad idea?